src/multinode_run.jl

```diff
@@ -1,193 +1,700 @@
+                  trust_region=NewtonTrustRegion(),
+                  timing=InferTiming())
+    # if this thread was late, move on quickly
+    if cbox.state[] == BoxReady
+        return
+    end
+
+    # only one thread does the partitioning and pre-allocation
+    lock(cbox.lock)
+    if cbox.state[] == BoxLoaded
+        cbox.sources_assignment = partition_box(nworkers, cbox.target_sources,
+                                                cbox.neighbor_map;
+                                                cyclades=cyclades,
+                                                batch_size=batch_size)
+
+        # configurations need to be persisted across calls to maximize! so
+        # that location constraints don't shift from their initial position
+        cbox.ea_vec = Vector{ElboArgs}(cbox.nsources)
+        cbox.cfg_vec = Vector{Config{DEFAULT_CHUNK,Float64}}(cbox.nsources)
+
+        # pre-allocate elbo args variational params
+        cbox.ts_vp = Dict{Int64,Array{Float64}}()
+        for ts in cbox.target_sources
+            cat = cbox.catalog[ts]
+            cbox.ts_vp[ts] = generic_init_source(cat.pos)
+        end
+
+        # update box state
+        cbox.state[] = BoxInitializing
+    end
+    unlock(cbox.lock)
+
+    # initialize elbo args for all sources
+    tic()
+    while cbox.state[] == BoxInitializing
+        lock(cbox.lock)
+        ts = cbox.curr_source
+        cbox.curr_source = cbox.curr_source + 1
         unlock(cbox.lock)
-        return true
+        if ts > cbox.nsources
+            if cbox.state[] == BoxInitializing
+                atomic_cas!(cbox.state, BoxInitializing, BoxReady)
+            end
+        else
+            #ntputs(grank(), threadid(), "initializing source $ts")
+            init_elboargs(ts, cbox.catalog, cbox.target_sources, cbox.neighbor_map,
+                          cbox.images, cbox.ts_vp, cbox.ea_vec, cbox.cfg_vec;
+                          use_fft=use_fft,
+                          trust_region=trust_region)
+        end
     end
+    timing.init_elbo = timing.init_elbo + toq()
+end
+
 
-    # thread function to process sources in boxes issued by Dtree
+"""
+Called by a thread to process a batch of sources in a box.
+"""
+function process_sources(sources::Vector{Int64}, cbox::BoxInfo;
+                         use_fft=false,
+                         within_batch_shuffling=true)
+    ntputs(grank(), threadid(), "in process_sources() with $sources")
+    ntputs(grank(), threadid(), "$(cbox.ea_vec)")
+    for s in sources
+        #ntputs(grank(), threadid(), "trying source $s, ea $(cbox.ea_vec[s])")
+        try
+            ea, cfg = cbox.ea_vec[s], cbox.cfg_vec[s]
+            if use_fft
+                f = FFTElboFunction(load_fsm_mat(ea, cbox.images; use_raw_psf=true))
+            else
+                f = DeterministicVI.elbo
+            end
+            ntputs(grank(), threadid(), "maximizing source $s")
+            NewtonMaximize.maximize!(f, ea, cfg)
```

Comment:
This call won't work because `maximize!` also needs `vp` based to it. Can you reuse the function in `joint_infer.jl` that calls NewtonMaximize, modifying that function as necessary? Those functions all are exercised by unit tests.